iree/hal/local/executable_library.h

```diff
@@ -82,33 +139,47 @@ typedef union {
-    __attribute__((align_value(16)));
-#endif  // MSVC
-
-typedef void* iree_hal_executable_binding_ptr_t;
+// Read-only per-dispatch state passed to each workgroup in a dispatch.
+typedef struct {
+  // Total workgroup count for the dispatch. This is sourced from either the
+  // original dispatch call (for iree_hal_command_buffer_dispatch) or the
+  // indirection buffer (for iree_hal_command_buffer_dispatch_indirect).
+  iree_hal_vec3_t workgroup_count;
+  // Workgroup size chosen for the dispatch. For compilation modes where the
+  // workgroup size is constant this may be ignored.
+  iree_hal_vec3_t workgroup_size;
+
+  // Total number of available 4 byte push constant values in |push_constants|.
+  size_t push_constant_count;
+  // |push_constant_count| values.
+  const uint32_t* push_constants;
+
+  // Total number of binding base pointers in |binding_ptrs| and
+  // |binding_lengths|. The set is packed densely based on which binidngs are
+  // used (known at compile-time).
+  size_t binding_count;
+  // Base pointers to each binding buffer.
+  void* const* binding_ptrs;
+  // The length of each binding in bytes, 1:1 with |binding_ptrs|.
+  const size_t* binding_lengths;
+
+  // Optional imported functions available for use within the executable.
+  const iree_hal_executable_import_table_v0_t* imports;
+} iree_hal_executable_dispatch_state_v0_t;
 
 // Function signature of exported executable entry points.
-// The same |state| is passed to all tiles in a dispatch, with other arguments
-// such as |workgroup_id| varying per-tile (counting to the |workgroup_count|).
-// Each tile represents |workgroup_size| local invocations in the global
-// |workgroup_count| grid.
+// The same |dispatch_state| is passed to all workgroups in a dispatch while
+// |workgroup_id| will vary for each workgroup.
 //
-// 0 or more push constants are available at |push_constants| with the count
-// being determined by the sidechannel information provided by the compiler.
-//
-// The |bindings| list is a dense set of pointers to I/O data with the count and
-// ordering determined by the compiler.
-typedef void (*iree_hal_executable_dispatch_v0_t)(
-    const iree_hal_executable_dispatch_state_v0_t* state,
-    const iree_hal_vec3_t* workgroup_id, const iree_hal_vec3_t* workgroup_size,
-    const iree_hal_vec3_t* workgroup_count,
-    const iree_hal_executable_push_constants_ptr_t push_constants,
-    const iree_hal_executable_binding_ptr_t* bindings);
+// Returns 0 on success and non-zero on failure. Failures will cause device loss
+// and should only be used to communicate serious issues that should abort all
+// execution within the current device. Buffer overflows are a good example of
+// a useful failure though the HAL does not mandate that all overflows are
+// caught and only that they are not harmful - clamping byte ranges and never
+// returning a failure is sufficient.
+typedef int (*iree_hal_executable_dispatch_v0_t)(
```

Comment:
+1, I think will be very useful (in the future) having executable to return some kind of status e.g block of code that verifies the dynamic shapes at runtime.